Add kmeans clustering

metasyn opened the pull request:
Hello :)

Here is my very first attempt at writing something more than a few lines long in nim... so I am positive there is a *lot* of room for improvement. Any feedback welcome, in the general coding style sense, and of course, the actual tensor manipulation / ML part of things.

Some preliminary benchmarks against sklearn (numpy + cython) says that this is a bit faster. Maybe after/if this gets merged, we can add the kmeans benchmark in a proper fashion.

I added some test cases for basic sanity checks. I also tried to incorporate the suggestions - e.g. I tried adding the `{.noInit.}` pragma and returning tuples instead of a transform flag.

Additionally, I tried to follow the kmeans++ algorithm more directly from the paper than from sklearn - although I did copy some parts of their implementation.

Here are a few parts that are glaringly wrong to me but I wasn't sure the right way to implement better with map or map inline, etc.
 
* excessive for-looping
* reformulating euclidean_distance to `sqrt ( x . x - 2 * y . y + y . y)` instead of `sqrt( (x - y) . (x - y) )`
* these kinds of obvious naivety:
```
var count = @[counts[i]].toTensor.astype(T)
centroids[i, _] = (totals[i] ./ count).reshape(1, n_cols)
```
or
```
distances ./ @[distances.sum].toTensor
```
* Tensor slicing / assignment - I have a feeling I'm doing it pretty ineffectually?

mratsim commented on tests/ml/test_clustering.nim:
```diff
@@ -0,0 +1,74 @@
+
+    check:
+      labels == transformed
+      labels == expected
```
Add a test that the initial data didn't change, it's easy to forget cloning ;).

metasyn replied:
Added this too!

metasyn commented on src/ml/clustering/kmeans.nim:
```diff
@@ -0,0 +1,268 @@
+  if not squared:
+    result = sqrt(result) 
+
+proc cumsum [T: SomeReal](p: Tensor[T]): Tensor[T] =
```
Oh this is cool. I added the todo.

metasyn commented on src/ml/clustering/kmeans.nim:
```diff
@@ -0,0 +1,268 @@
+  assert p.rank == 1
+  assert p.shape[1] == 0
+  var n_rows = p.shape[0]
+  result = p
```
Ah, yes, I had noticed this too, and had considered calling in cumsum_inplace or something to make that obvious - I changed it to using p.clone() but it doesn't affect the result in this case, but better to be clear I think, too :)

metasyn commented on src/ml/clustering/kmeans.nim:
```diff
@@ -0,0 +1,268 @@
+proc get_candidates [T: SomeReal](n: int, distances: Tensor[T]): Tensor[int] =
+  ## Sample candidates with probability weighted by the distances
+  var
+    probs = cumsum(distances ./ @[distances.sum].toTensor)
```
I tired removing the additional tensor related code for the denominator, but it fails because of a type mismatch:

e.g. 

```
probs = cumsum(distances ./ distances.sum)

proc `./`[T: SomeInteger](a, b: Tensor[T]): Tensor[T]
  first type mismatch at position: 1
  required type: Tensor[./.T]
  but expression 'distances' is of type: Tensor[system.float64]
proc `./`[T: SomeFloat](val: T; t: Tensor[T]): Tensor[T]
  first type mismatch at position: 1
  required type: T: SomeFloat
  but expression 'distances' is of type: Tensor[system.float64]
proc `./`[T: SomeInteger](val: T; t: Tensor[T]): Tensor[T]
  first type mismatch at position: 1
  required type: T: SomeInteger
  but expression 'distances' is of type: Tensor[system.float64]
proc `./`[T: SomeFloat](a, b: Tensor[T]): Tensor[T]
  first type mismatch at position: 2
  required type: Tensor[./.T]
  but expression 'sum(distances)' is of type: float64

expression: distances ./ sum(distances)
```

I think `distances ./ distances` but that would just reduce each value to 1. I switched the var to let.

metasyn commented:
Ok I think I addressed most of the comments, except this line:

```
let probs = cumsum(distances ./ @[distances.sum].toTensor)
```

which I am happy to change if there is a better way.

Thanks for the review!

metasyn commented:
Also, this was what I was doing for a bench mark:
```
from numpy import load 
from sklearn.cluster import KMeans
iris = load("/Users/aljohnson/desktop/iris.npy")
c = KMeans(n_clusters=3, n_init=10000, random_state=42)
c.fit_transform(iris)
```
and timing naively like:
```
➜ time python kmeans.py 

real	0m18.564s
user	0m17.768s
sys	0m0.319s
```

compared to Nim (`nim c -d:release kmeans.nim`):

```
import arraymancer
var iris = read_npy[float32]("/Users/aljohnson/Desktop/iris.npy")
var (labels, centroids, inertia) = iris.kmeans(3, n_init=10000, seed=42)
```
which gives:
```
➜ time ./kmeans

real	0m6.832s
user	0m6.792s
sys	0m0.017s
```

metasyn commented on src/ml/clustering/kmeans.nim:
```diff
@@ -0,0 +1,268 @@
+  ## sqrt( (u - v) dot (u - v) )
+  ## which can be reformulated
+  ## sqrt((u dot u) - 2 * (u dot v) + (v dot v))
+  var u_v = (u .- v).reshape(u.shape[1])
```
I checked on the reshape, looks like dot is not defined for tensors greater than rank 1, so this reshape is force rank of 1. 

Removing it produces:
```
kmeans.nim(23)           euclidean_distance
operators_blas_l1.nim(28) dot
p_checks.nim(109)        check_dot_prod

    Unhandled exception: Dot product is only supported for vectors (tensors of rank 1) [ValueError]
```

mratsim commented:
Excellent, thank you very much

mratsim commented:
By the way, you might want to test with `-d:openmp`, by default matrix multiplication and matrix-vector multiplication are parallelized (by the BLAS) but everything else needs the openmp flag to be parallel.

mratsim commented:
I noticed that some broadcasted division `./` were missing, fixed by https://github.com/mratsim/Arraymancer/commit/afc0c213316b4d98002fbb118551a4de9b67e217, I update the `cumsum(distances ./ @[distances.sum].toTensor)` as well
